rust/lance-core/src/io/writer/statistics.rs

```diff
@@ -49,75 +55,385 @@ pub struct StatisticsRow {
+    // If all bytes are already maximal, no need to truncate
+
+    Some(&data[0..length])
+}
+
+/// Try and increment the bytes from right to left.
+///
+/// Returns `None` if all bytes are set to `u8::MAX`.
+fn increment(mut data: Vec<u8>) -> Option<Vec<u8>> {
+    for byte in data.iter_mut().rev() {
+        let (incremented, overflow) = byte.overflowing_add(1);
+        *byte = incremented;
+
+        if !overflow {
+            return Some(data);
+        }
+    }
+
+    None
+}
+
+fn get_string_statistics<T: OffsetSizeTrait>(arrays: &[&ArrayRef]) -> StatisticsRow {
+    let mut min_value: Option<&str> = None;
+    let mut max_value: Option<&str> = None;
+    let mut null_count: i64 = 0;
+
+    let array_iterator = arrays.iter().map(|x| x.as_string::<T>());
+
+    for array in array_iterator {
+        null_count += array.null_count() as i64;
+        if array.null_count() == array.len() {
+            continue;
+        }
+
+        array.iter().for_each(|value| {
+            if let Some(mut val) = value {
+                for i in (BINARY_PREFIX_LENGTH..val.len() + 4).rev() {
+                    if val.is_char_boundary(i) {
+                        val = &val[..i];
+                        break;
+                    }
+                }
+
+                if let Some(v) = min_value {
+                    if let Some(Ordering::Less) = val[..].partial_cmp(v) {
+                        min_value = Some(val);
+                    }
+                } else {
+                    min_value = Some(val);
+                }
+
+                if let Some(v) = max_value {
+                    if let Some(Ordering::Greater) = val.partial_cmp(v) {
+                        max_value = Some(val);
+                    }
+                } else {
+                    max_value = Some(val);
+                }
+            }
+        });
+    }
+
+    if let Some(v) = min_value {
+        if v.len() > BINARY_PREFIX_LENGTH {
+            min_value = truncate_utf8(v, BINARY_PREFIX_LENGTH);
+        }
+    }
+
+    let max_value_bound: Vec<u8>;
+    if let Some(v) = max_value {
+        if v.len() > BINARY_PREFIX_LENGTH {
+            max_value = truncate_utf8(v, BINARY_PREFIX_LENGTH);
+            max_value_bound = increment_utf8(max_value.unwrap().as_bytes().to_vec()).unwrap();
+            max_value = Some(str::from_utf8(&max_value_bound).unwrap());
+        }
+    }
+
+    let min_value = min_value.map(|x| x.to_string());
+    let max_value = max_value.map(|x| x.to_string());
+
+    match arrays[0].data_type() {
+        DataType::Utf8 => StatisticsRow {
+            null_count,
+            min_value: ScalarValue::Utf8(min_value),
+            max_value: ScalarValue::Utf8(max_value),
+        },
+        DataType::LargeUtf8 => StatisticsRow {
+            null_count,
+            min_value: ScalarValue::LargeUtf8(min_value),
+            max_value: ScalarValue::LargeUtf8(max_value),
+        },
+        _ => {
+            todo!()
```

Comment:
nit: I think you mean `unreachable!()` here, right? Also if you are leaving a `todo!()` call in a final PR, could you at least write a message in it? (Such as `todo!("description of what I will do later")`) This applies throughout this file.

Reply:
Yes! Updated. Left two `todo!()` for the `Decimal256`.